### ISSUE TITLE
Naming Suggestions

### ISSUE THREAD

**Issue by srush** [google/objax#14]
This library will likely be quite popular among those used to PyTorch in the research community, as there is a demand for a standard NN interface to jax. Already the module/optimizer design is very familiar to those who have used pytorch, and the developer clearly was inspired by this design structure. 

However, there is one aspect that is challenging to understand. In PyTorch (and in Autodiff), variable is synonymous with "value that you take a gradient wrt". In Objax, variable has a different meaning both in `TrainVar` (`Parameter` in pytorch) and `StateVar` (`Buffer` in pytorch). TrainVar is a variable, whereas StateVar is not. Furthermore `Train` is confusing as terminology, since it only applies during model fitting, as opposed to during inference. 

Given this similarity, it seems like a natural change would be to refer to TrainVar => Parameter and StateVar => State or Buffer. These names avoid the ambiguity of variable, dictate what they do not how they are used, and make the library even more familiar and easy to use for those used to PyTorch.

**Comment by AlexeyKurakin** [comment]
Thanks for your feedback.

Objax borrows inspiration from many frameworks: JAX, TF, PyTorch and Keras. Thus users may see some similarities between Objax API and APIs of other frameworks. But to be perfectly clear, Objax is not meant as a reimplementation of PyTorch on top of JAX. We (authors of Objax) decided to not import decisions from other frameworks directly into Objax. Instead we decided to name and design things according to their own merit in our eyes.

On the topic of the word Parameter (from PyTorch) to be used instead of TrainVar, it should be noted that the word parameter is ambiguous between maths, programming languages and machine learning. In math, we refer to a function f(x; \theta) as a function `f` of x parameterized over \theta. Here `\theta` refers to an attribute of a function (or class). In programming languages, however, the word parameter is used to describe an input to a function: `def f(x)` would be described as the function `f` taking a parameter `x`.
Finally, in the context of machine learning, parameter has yet another ambiguous interpretation: “hyper-parameter”.

Due to these ambiguities, we don’t think the word parameter is well suited for clarity considerations.

To share more about our thought process, here’s how we converged to a naming in discussions among ourselves when writing Objax:
Generally in programming, variable (shorthand `Var`) is any value which can change. In this sense both `TrainVar` and `StateVar` are variables. `TrainVar` changes by optimizer, `StateVar` changes by batch norm update rules (or any other rules where it’s used).
`Train` is a shorthand for trainable, so `TrainVar` means a variable which encapsulates “trainable” value (the one changed by optimizer). `StateVar` encapsulates state which is not trainable (i.e. not being changed by optimizer).

We understand that this has a clear machine learning flavor and that’s an assumed choice: Objax’s audience is the machine learning community.

**Comment by srush** [comment]
Okay, I understand this is your library, but let me make one more pitch. 

Non-standard names make things much harder to learn for students. Everyone working on neural networks knows what a parameter is, and standardization is wonderful. If Objax is just PyTorch for jax, that is great! It doesn't have to be a new creative thing, I would use it in my class tomorrow. In fact that is exactly how it was pitched by the author: 

"Objax, a new high-level JAX API with a PyTorch-like interface!"

(Closing the issue)

**Comment by david-berthelot** [comment]
But why not use PyTorch then, what's the value in porting it verbatim to JAX? The original is always better than the copy.

We wrote Objax because there are lot of things we saw missing in other frameworks, and for the things that were good we didn't reinvent the wheel just for the sake it either; we simply kept them. We thought hard about inheriting decisions we don't agree with from other frameworks, but we decided against it since it perpetuates them.

Btw we only started this framework, anyone can fork it to make it what they feel it should be if some decision seems unacceptable to them. And you could also keep the issue open to gather a richer set of opinions if you want to, it's not like our views can't be changed.

**Comment by srush** [comment]
People love the pytorch API, it's super well run, standardized, and accessible. Students find it amazing to use. We teach all our courses in it.

But JAX is faster, optimized, and supports lots of interesting higher-order operations. But the APIs are a mess! We just need a great frontend to this library.

**Comment by david-berthelot** [comment]
Thanks for the context, I understand better your suggestions now. Well one thing to keep in mind is to question whether you like a decision out of habit or because it makes sense. We asked ourselves this question a lot, esp. since we matured and refined Objax for 6 months while using it for our own research. We thought, even overthought, some decisions a lot. In the end, we have to do what we think is right! 

Thanks for the feedback.
